pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,225 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: jefftk@google.com (Jeff Kaufman)
+
+#include "pagespeed/system/controller_manager.h"
+
+#include <signal.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstddef>
+#include <cstdlib>
+
+#include "base/logging.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/system/system_rewrite_options.h"
+
+namespace net_instaweb {
+
+
+ControllerManager::ProcessDeathWatcherThread::ProcessDeathWatcherThread(
+    ThreadSystem* thread_system,
+    int controller_read_fd,
+    MessageHandler* handler) : Thread(thread_system,
+                                      "process death watcher",
+                                      ThreadSystem::kDetached),
+                              handler_(handler),
+                              controller_read_fd_(controller_read_fd) {
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(kInfo, "Watching the root process to exit if it does.");
+
+  ssize_t status;
+  char buf[1];
+  do {
+    status = read(controller_read_fd_, buf, 1);
+  } while (status == -1 && (errno == EINTR ||
+                            errno == EAGAIN));
+  if (status == -1) {
```

Comment:
The difference between pipe error, root process exiting, and root process starting a new controller seems like it would be useful to us for debugging in their field if something is wrong.  I'd rather keep these?

If you don't like the verbosity I guess we could log "lost contact with root process, exiting (status=%d, errno=%d)" but I think the way I have it is more likely to make sense to someone trying to debug this later.